lib/CL/devices/hsa/pocl-hsa.c

```diff
@@ -496,6 +681,145 @@ pocl_hsa_run
 }
 
 void
+pocl_hsa_copy (void *data, const void *src_ptr, size_t src_offset,
+                 void *__restrict__ dst_ptr, size_t dst_offset, size_t cb)
+{
+  if (src_ptr == dst_ptr)
+    return;
+
+  memcpy ((char*)dst_ptr + dst_offset, (char*)src_ptr + src_offset, cb);
+}
+
+void
+pocl_hsa_copy_rect (void *data,
```

Comment:
If this is identical to the implementation of 'basic' device's, you should not copy paste the code here, but just use the pocl_basic_copy_rect directly by assigning it in the function pointer table.